### DISCUSSION TITLE
Can we filter comparing one column with another?

### DISCUSSION THREAD

**Discussion by mattscooter** [mikro-orm/mikro-orm#4218] [Q&A] [answered]
Is it possible to build a find filter where we compare one column with another?

For example I have an entity thing:

```
@Entity
export class Thing 
{
  @PrimaryKey()
  id: string;

  @Property()
  maxScore: number = 100;
  
  @Property()
  currentScore: number;
}
```

I want to find all Things where the currentScore value is less than the maxScore.  I know this is easy to do in a Query Builder but I was hoping to be able to do this in the standard repository find.  

I was hoping it would be as simple as something like:

	orm.em.find(Thing, {currentScore: { $lt: { $field: 'maxScore' } } }); 

Obviously this doesn't work because $field isn't a valid option here but it feels like it should be that simple!  ;)

I've been through the documentation a few times now but can't see any example of how to do this. Have I somehow missed this or is it not possible? Do I just have to use the QB for these types of queries?

**Comment by B4nan** [comment]
You can use a raw sql fragment for this. 

https://mikro-orm.io/docs/entity-manager#using-custom-sql-fragments

**Reply by simonbrunel** [reply]
@B4nan I'm trying to implement a similar query but can't figure how to translate @mattscooter example to SQL fragment:

```ts
{ currentScore: { $lt: { $field: 'maxScore' } } }
```

Can you please write how to use it with the `$lt` operator?

BTW, I like @mattscooter suggestion of the `$field` operator since it adds type checking to the field name (`maxScore` will be converted to `max_score` and if the field name change, it will stop compiling, contrary to using raw SQL).

**Reply by B4nan** [reply, marked as answer]
Actually, that link is not really describing what you were asking for, for this `em.raw()` helper is needed:

```ts
await em.find(Thing, { currentScore: { $lt: em.raw('maxScore') } }); 
```

Note that it needs to be a column name, so maybe you want `em.raw('max_score')` instead. You could combine this with 

In v6 this `raw` helper is a regular function (not an instance method): https://github.com/mikro-orm/mikro-orm/blob/v6/docs/docs/raw-queries.md

Indeed, the `$field` proposal sounds quite nice.

**Reply by simonbrunel** [reply]
_Edit: I didn't see your previous reply before submitting this comment._

After a few more attempts, it seems to be:

```ts
{ [expr("current_score")]: { $lt: em.raw("max_score") } } }
```

Would still be great to have type checking for such simple use cases.

Also, not sure how to use `expr()` and `raw()` with nested fields:

```ts
{ currentScore: { $lt: { $field: 'group.maxScore'} }
```

In this case, the `group` table needs to be left joined (if not already) and we will need to know about the alias (i.e. `"g0"."max_score"`).

**Reply by B4nan** [reply]
v6 is already addressing the problem of aliasing, as the new `raw` helper (which is a replacement for all of `expr/em.raw/qb.raw`) accepts a callback signature too - where you have the current alias in the callback parameter.

For more complex queries you can always use query builder, that way you can control joins and aliasing explicitly.

**Reply by mattscooter** [reply]
Thanks for providing the way to achieve this without having to resort to using a query builder! I'll be sure to use this going forward.
